### PR TITLE
Configurable verbosity for turnkey status

### LINKED ISSUE #113: Configurable verbosity in Analyzer Status
Problem: 
- the Analyzer orchestrates the analysis, build, and benchmarking of all the models in one file. It is typically invoked via the Files API, which in turn is typically invoked via the CLI.
- Analyzer has a module, `analyze.status`, which updates the terminal with progress. This status takes over the whole terminal, clearing it and updating it with information for all models across all files at each step.
- This status update system makes the CLI function like an application, which is great for demos and low numbers of models and files.
- It is not suitable for large numbers of models/files, or for direct use of the API, because it repeatedly prints full information for every model to the screen. This results in O(N^2) prints, which is not good from an API logging perspective.

Proposed solution: add verbosity modes to the status module that allow it to be configured for the following scenarios:
1. App mode: the way status currently works
2. Mass-evaluation mode: print status for each model exactly once, as that status becomes available. Don't clear the terminal. 
   - First print:  Model identified by analysis
   - Second print: Build status of model
   - Third print: Benchmark status of model
   - Also: show a TQDM-style status bar as we iterate over the files
3. Quiet mode: print minimal information (e.g., only the result requested by an API)

cc @danielholanda

### PR DESCRIPTION
**Pull request by jeremyfowers**
Closes #113 by adding configurable verbosity settings
Closes https://github.com/onnx/turnkeyml/issues/25, an old bug. It turns out we were mixing up invocation hash vs. model hash. I have fixed the bug and clarified the code.

# Major Changes

## Configurable Verbosity 

Four new verbosity levels:
  - `dynamic`: turnkey's traditional default mode, with slightly lower verbosity (no more hash or model class)
  - `static`: improved version of `TURNKEY_DEBUG=True`. Prints each piece of info about each invocation exactly once. Does not clear the terminal.
  - `auto`: automatically select one of the above.

`auto` is the default for the CLI, and will set the verbosity level automatically based on the following policies:
- CLI commands with 4 or fewer input files: `dynamic`
- CLI commands with 5 or more input files, CLI in process isolation mode: `static`
   - Note: try `turnkey examples/cli/scripts/*.py --process-isolation` on main. It is not good!

`simple` is the default for the API.

Additionally, users can override the `auto` policy with a new env var, `TURNKEY_VERBOSITY`.

Finally, the `TURNKEY_DEBUG` env var has been removed. Tests have been updating to use `TURNKEY_VERBOSITY=static` instead.

## Progress Bar

A progress bar is displayed whenever >4 files are being evaluated.

## Under the Hood

The class `turnkey.analyze.util.UniqueInvocationInfo` is now capable of printing its own status. We no longer rely on `print_invocation()` for this.

Details:
- UniqueInvocationInfo status printing is now stateful. A class member, `skip`, indicates which status fields should be skipped. This is useful for:  
  - Low verbosity modes, such as `app_low` mode
  - Keeping track of which fields have already been printed, to avoid re-printing them (used in `simple` mode)
- Each line of status is printed with its own sub-function within `UniqueInvocationInfo.print()`. This makes it easy to add new fields, reorder the fields, skip fields, etc.

Additionally, there is a new base class, `BasicInfo`, that is inherited by both `ModelInfo` and `UniqueInvocationInfo`. This refactor was needed to avoid a circular import where `UniqueInvocationInfo.print()` wanted to access `ModelInfo` data, but `ModelInfo` had the `UniqueInvocationInfo` as a member. 

Finally, everything related to status, including  `ModelInfo` and `UniqueInvocationInfo`, have been moved to the `status` module.

# Testing Methodology

Since this is a UI/UX change, it is only partially covered by our programmatic tests. In order to assure quality, I ran through all of the CLI and API tutorials and made sure that the contents of the terminal matched my expectations. This process was made easier because the tutorials themselves had copies of the old terminal contents, so I could make sure that they literally matched.

Note: one of the tutorials had a bug, which I fixed in this PR.

# Demo

## App Mode

App mode is completely unchanged from how `turnkey` used to work.

```
jefowers@XSJ-JEFOWERS-L2:~/turnkeyml/examples/cli$ turnkey scripts/*.py
```

image

## App-Low Mode

This is our new demo mode. Note that the `hash` and `model type` fields have been omitted.

```
jefowers@XSJ-JEFOWERS-L2:~/turnkeyml/examples/cli$ export TURNKEY_VERBOSITY=app_low
jefowers@XSJ-JEFOWERS-L2:~/turnkeyml/examples/cli$ turnkey scripts/*.py
```

image

## Simple Mode

This will be used for mass-benchmarking and working with the API.

```
jefowers@XSJ-JEFOWERS-L2:~/turnkeyml/examples/cli$ export TURNKEY_VERBOSITY=simple
jefowers@XSJ-JEFOWERS-L2:~/turnkeyml/examples/cli$ turnkey scripts/*.py
```

```
(tkml) (miniconda3) jefowers@XSJ-JEFOWERS-L2:~/turnkeyml/examples/cli$ turnkey scripts/*.py
  0%|                                                                                                                           | 0/4 [00:00<?, ?it/s]
Info: Running turnkey on scripts/hello_world.py
hello_world.py:
        pytorch_model (executed 1x)
                Model Type:     Pytorch (torch.nn.Module)
                Class:          SmallModel (<class '__main__.SmallModel'>)
                Location:       /home/jefowers/turnkeyml/examples/cli/scripts/hello_world.py, line 30
                Parameters:     60 (240.0 B)
                Input Shape:    'x': (11,)
                Hash:           2a1e15ca
                Build dir:       /home/jefowers/.cache/turnkey/hello_world_turnkey_24adc42f
                Status:         Computing...


Capturing statistics in turnkey_stats.yaml under evaluation ID: 1458bd2f



Building "hello_world_turnkey_24adc42f"
    ✓ Exporting PyTorch to ONNX   
    ✓ Optimizing ONNX file   

Woohoo! Saved to ~/.cache/turnkey/hello_world_turnkey_24adc42f

                Status:         Successfully benchmarked on AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics (ort v1.15.1) 
                                Mean Latency:   0.010   milliseconds (ms)
                                Throughput:     101060.0        inferences per second (IPS)



pytorch_outputs: tensor([ 0.3808, -0.0187, -0.1056,  0.1954,  0.1319], grad_fn=<ViewBackward0>)
 25%|████████████████████████████▊                                                                                      | 1/4 [00:06<00:19,  6.59s/it]
Info: Running turnkey on scripts/max_depth.py
max_depth.py:
        pytorch_model (executed 1x)
                Model Type:     Pytorch (torch.nn.Module)
                Class:          TwoLayerModel (<class '__main__.TwoLayerModel'>)
                Location:       /home/jefowers/turnkeyml/examples/cli/scripts/max_depth.py, line 42
                Parameters:     85 (340.0 B)
                Input Shape:    'x': (10,)
                Hash:           80b93950
                Build dir:       /home/jefowers/.cache/turnkey/max_depth_turnkey_7ebbe8c4
                Status:         Computing...


Capturing statistics in turnkey_stats.yaml under evaluation ID: 97a042cc



Building "max_depth_turnkey_7ebbe8c4"
    ✓ Exporting PyTorch to ONNX   
    ✓ Optimizing ONNX file   

Woohoo! Saved to ~/.cache/turnkey/max_depth_turnkey_7ebbe8c4

                Status:         Successfully benchmarked on AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics (ort v1.15.1) 
                                Mean Latency:   0.010   milliseconds (ms)
                                Throughput:     103169.8        inferences per second (IPS)



pytorch_outputs: tensor([-0.5411,  0.3897,  0.3323, -0.5417,  0.0850], grad_fn=<ViewBackward0>)
 50%|█████████████████████████████████████████████████████████▌                                                         | 2/4 [00:12<00:12,  6.29s/it]
Info: Running turnkey on scripts/multiple_invocations.py
multiple_invocations.py:
        pytorch_model (executed 1x)
                Model Type:     Pytorch (torch.nn.Module)
                Class:          SmallModel (<class '__main__.SmallModel'>)
                Location:       /home/jefowers/turnkeyml/examples/cli/scripts/multiple_invocations.py, line 41
                Parameters:     65 (260.0 B)
                Input Shape:    'x': (1, 12)
                Hash:           d0b6b257
                Build dir:       /home/jefowers/.cache/turnkey/multiple_invocations_ef80819e
                Status:         Computing...


Capturing statistics in turnkey_stats.yaml under evaluation ID: b1a86cb1



Building "multiple_invocations_ef80819e"
    ✓ Exporting PyTorch to ONNX   
    ✓ Optimizing ONNX file   

Woohoo! Saved to ~/.cache/turnkey/multiple_invocations_ef80819e

                Status:         Successfully benchmarked on AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics (ort v1.15.1) 
                                Mean Latency:   0.010   milliseconds (ms)
                                Throughput:     103675.0        inferences per second (IPS)




multiple_invocations.py:
        pytorch_model
                Model Type:     Pytorch (torch.nn.Module)
                Class:          SmallModel (<class '__main__.SmallModel'>)
                Location:       /home/jefowers/turnkeyml/examples/cli/scripts/multiple_invocations.py, line 41
                Parameters:     65 (260.0 B)

                With input shape 1 (executed 1x)
                Input Shape:    'x': (2, 12)
                Hash:           d0b6b257
                Build dir:       /home/jefowers/.cache/turnkey/multiple_invocations_d5267eaf
                Status:         Computing...


Capturing statistics in turnkey_stats.yaml under evaluation ID: b1a86cb1



Building "multiple_invocations_d5267eaf"
    ✓ Exporting PyTorch to ONNX   
    ✓ Optimizing ONNX file   

Woohoo! Saved to ~/.cache/turnkey/multiple_invocations_d5267eaf

                Status:         Successfully benchmarked on AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics (ort v1.15.1) 
                                Mean Latency:   0.010   milliseconds (ms)
                                Throughput:     95358.1 inferences per second (IPS)



 75%|██████████████████████████████████████████████████████████████████████████████████████▎                            | 3/4 [00:24<00:08,  8.94s/it]
Info: Running turnkey on scripts/two_models.py
two_models.py:
        pytorch_model (executed 1x)
                Model Type:     Pytorch (torch.nn.Module)
                Class:          SmallModel (<class '__main__.SmallModel'>)
                Location:       /home/jefowers/turnkeyml/examples/cli/scripts/two_models.py, line 33
                Parameters:     55 (220.0 B)
                Input Shape:    'x': (10,)
                Hash:           479b1332
                Build dir:       /home/jefowers/.cache/turnkey/two_models_turnkey_f93db89f
                Status:         Computing...


Capturing statistics in turnkey_stats.yaml under evaluation ID: 87a7d9f5



Building "two_models_turnkey_f93db89f"
    ✓ Exporting PyTorch to ONNX   
    ✓ Optimizing ONNX file   

Woohoo! Saved to ~/.cache/turnkey/two_models_turnkey_f93db89f

                Status:         Successfully benchmarked on AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics (ort v1.15.1) 
                                Mean Latency:   0.009   milliseconds (ms)
                                Throughput:     107771.7        inferences per second (IPS)



two_models.py:
        another_pytorch_model (executed 1x)
                Model Type:     Pytorch (torch.nn.Module)
                Class:          SmallModel (<class '__main__.SmallModel'>)
                Location:       /home/jefowers/turnkeyml/examples/cli/scripts/two_models.py, line 41
                Parameters:     510 (1.99 KB)
                Input Shape:    'x': (50,)
                Hash:           215ca1e3
                Build dir:       /home/jefowers/.cache/turnkey/two_models_turnkey_c69b7dea
                Status:         Computing...


Capturing statistics in turnkey_stats.yaml under evaluation ID: 87a7d9f5



Building "two_models_turnkey_c69b7dea"
    ✓ Exporting PyTorch to ONNX   
    ✓ Optimizing ONNX file   

Woohoo! Saved to ~/.cache/turnkey/two_models_turnkey_c69b7dea

                Status:         Successfully benchmarked on AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics (ort v1.15.1) 
                                Mean Latency:   0.004   milliseconds (ms)
                                Throughput:     225600.7        inferences per second (IPS)



pytorch_outputs: tensor([ 0.3628,  0.0489,  0.2952,  0.0022, -0.0161], grad_fn=<ViewBackward0>)
more_pytorch_outputs: tensor([-0.1198, -0.5344, -0.1920, -0.1565,  0.2279,  0.6915,  0.8540, -0.2481,
         0.0616, -0.4501], grad_fn=<ViewBackward0>)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:36<00:00,  9.23s/it]

Woohoo! The 'benchmark' command is complete.
```

### CODE REVIEW

File: examples/cli/discovery.md [hunk 1]
```diff
@@ -54,12 +54,6 @@ If you want to benchmark an entire corpus of models, but you don't want to call
 
 For example, the command:
 
-```
-turnkey scripts/hello_world.py scripts/two_models.py scripts/max_depth.py
```
Comment (jeremyfowers) [inline examples/cli/discovery.md:58]: Removed because this wasn't equivalent to the command below.

File: examples/cli/discovery.md [hunk 2]
```diff
@@ -117,6 +161,7 @@ To achieve the same result as above, you may simply call `turnkey selected_model
 hello_world.py
 two_models.py
 max_depth.py
+multiple_invocations.py
```
Comment (jeremyfowers) [inline examples/cli/discovery.md:164]: Added to make this equivalent to the command above.

File: examples/cli/discovery.md [hunk 3]
```diff
@@ -73,42 +67,92 @@ Models discovered during profiling:
...
+                Hash:           c69b7dea
+                Build dir:       /home/jefowers/.cache/turnkey/two_models_turnkey_c69b7dea
+                Status:         Successfully benchmarked on AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics (ort v1.15.1) 
+                                Mean Latency:   0.011   milliseconds (ms)
+                                Throughput:     92059.3 inferences per second (IPS)
+
+
+pytorch_outputs: tensor([ 0.3628,  0.0489,  0.2952,  0.0022, -0.0161], grad_fn=<ViewBackward0>)
+more_pytorch_outputs: tensor([-0.1198, -0.5344, -0.1920, -0.1565,  0.2279,  0.6915,  0.8540, -0.2481,
+         0.0616, -0.4501], grad_fn=<ViewBackward0>)
```
Comment (jeremyfowers) [inline examples/cli/discovery.md:150]: Replaced this printout because the second model in `two_models.py` was omitted,

File: examples/cli/discovery.md [hunk 4]
```diff
@@ -192,7 +237,7 @@ You can leverage the model hashes feature of `turnkey` to filter which models ar
 If you wanted to only build and benchmark `another_pytorch_model`, you could use this command, which filters `two_models.py` with the hash `215ca1e3`:
 
 ```
-turnkey benchmark scripts/two_models.py::215ca1e3
+turnkey benchmark scripts/two_models.py::c69b7dea
```
Comment (jeremyfowers) [inline examples/cli/discovery.md:240]: The prior hash got stale at some point and needed to be replaced.

At some point we might want to auto-generate the docs to ensure correctness of printouts and variables ;)

File: examples/cli/scripts/multiple_invocations.py [hunk 5]
```diff
@@ -32,7 +33,7 @@ def forward(self, x):
 
 
 # Instantiate model and generate inputs
-input_features = 11
+input_features = 12
```
Comment (jeremyfowers) [inline examples/cli/scripts/multiple_invocations.py:36]: This was a bugfix; the models were identical across two of the example scripts, which made `turnkey scripts/*.py` malfunction.

File: src/turnkeyml/analyze/status.py [hunk 6]
```diff
@@ -78,149 +129,3 @@ def recursive_print(
                     parent_invocation_hash=invocation_hash,
                     script_names_visited=script_names_visited,
                 )
-
-
-def _pretty_print_key(key: str) -> str:
```
Comment (jeremyfowers) [inline src/turnkeyml/analyze/status.py:83]: Moved to util.py (no change)

File: src/turnkeyml/analyze/status.py [hunk 7]
```diff
@@ -78,149 +129,3 @@ def recursive_print(
...
-    result = key.split("_")
-    result = [word.capitalize() for word in result]
-    result = " ".join(result)
-    return result
-
-
-def parameters_to_size(parameters: int, byte_per_parameter: int = 4) -> str:
```
Comment (jeremyfowers) [inline src/turnkeyml/analyze/status.py:90]: Moved to util.py (no change)

File: src/turnkeyml/analyze/status.py [hunk 8]
```diff
@@ -78,149 +129,3 @@ def recursive_print(
...
-    if size_bytes == 0:
-        return "0B"
-    size_name = ("B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB")
-    i = int(math.floor(math.log(size_bytes, 1024)))
-    p = math.pow(1024, i)
-    s = round(size_bytes / p, 2)
-    return "%s %s" % (s, size_name[i])
-
-
-def print_invocation(
```
Comment (jeremyfowers) [inline src/turnkeyml/analyze/status.py:101]: Replaced by `UniqueInvocationInfo.print()`

File: src/turnkeyml/analyze/util.py [hunk 9]
```diff
@@ -17,15 +19,77 @@ class AnalysisException(Exception):
...
 
 
+def _pretty_print_key(key: str) -> str:
+    result = key.split("_")
+    result = [word.capitalize() for word in result]
+    result = " ".join(result)
+    return result
+
+
+def parameters_to_size(parameters: int, byte_per_parameter: int = 4) -> str:
```
Comment (jeremyfowers) [inline src/turnkeyml/analyze/util.py:29]: Moved from status.py

File: src/turnkeyml/analyze/script.py [hunk 10]
```diff
@@ -749,7 +761,17 @@ def forward_spy(*args, **kwargs):
...
+                        script_name=model_info.script_name,
+                        file=model_info.file,
+                        line=model_info.line,
+                        params=model_info.params,
+                        depth=model_info.depth,
+                        build_model=model_info.build_model,
+                        model_type=model_info.model_type,
+                        model_class=type(model_info.model),
+                        invocation_hash=invocation_hash,
+                        hash=model_info.hash,
```
Comment (jeremyfowers) [inline src/turnkeyml/analyze/script.py:774]: If anyone knows a better way to bulk-copy this info from `model_info` let me know...

File: src/turnkeyml/analyze/util.py [hunk 11]
```diff
@@ -17,15 +19,77 @@ class AnalysisException(Exception):
...
+    line: int = 0
+    params: int = 0
+    depth: int = 0
+    parent_hash: Union[str, None] = None
+    build_model: bool = False
+    model_type: build.ModelType = build.ModelType.PYTORCH
+    model_class: type = None
+    # This is the "model hash", not to be confused with the
+    # "invocation hash"
+    hash: Union[str, None] = None
```
Comment (jeremyfowers) [inline src/turnkeyml/analyze/util.py:60]: This confusion was the cause of #25

**Review by danielholanda — APPROVED** [review]
